go/caddy/log-tailer/logmeta.go

```diff
@@ -0,0 +1,181 @@
+package logtailer
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	logDebug = "debug"
+	logInfo  = "info"
+	logWarn  = "warn"
+	logError = "error"
+)
+
+type LogMetaEngine struct {
+	SyslogMap           map[string]*LogMetaExtractor
+	GlobalExtractionRe  *regexp.Regexp
+	TimestampPos        int
+	HostnamePos         int
+	SyslogNamePos       int
+	LogWithoutPrefixPos int
+}
+
+func reExtractor(re *regexp.Regexp, capturePosition int) func(string, string) string {
+	return func(syslogName string, log string) string {
+		if m := re.FindAllStringSubmatch(log, -1); m != nil {
+			return m[0][capturePosition]
+		} else {
+			return ""
+		}
+	}
+}
+
+var logLevelImplicitDebug = func(string, string) string { return logDebug }
+var logLevelImplicitInfo = func(string, string) string { return logInfo }
+var logLevelImplicitWarn = func(string, string) string { return logWarn }
+var logLevelImplicitError = func(string, string) string { return logError }
+
+var errorRegexp = regexp.MustCompile(`(?i).*error.*`)
+var freeradiusMetaExtractor = LogMetaExtractor{
+	LogLevelExtractor: func(syslogName, log string) string {
+		if errorRegexp.MatchString(log) {
+			return logError
+		} else {
+			return logInfo
+		}
+	},
+	ProcessNameExtractor: func(syslogName string, log string) string {
+		return "radiusd"
+	},
+}
+
+var golangMetaExtractor = LogMetaExtractor{
+	LogLevelExtractor: reExtractor(regexp.MustCompile(`lvl=([a-z]+)`), 1),
+	LogLevelNormalizer: func(level string) string {
+		return map[string]string{
+			"dbug": logDebug,
+			"info": logInfo,
+			"warn": logWarn,
+			"eror": logError,
+		}[level]
+	},
+	ProcessNameExtractor: func(syslogName, log string) string {
+		return syslogName
+	},
+}
+
+var log4perlMetaExtractor = LogMetaExtractor{
+	LogLevelExtractor: reExtractor(regexp.MustCompile(`^(\S+\s+){6}([A-Z]+)`), 2),
+	LogLevelNormalizer: func(level string) string {
+		return strings.ToLower(level)
+	},
+	ProcessNameExtractor: reExtractor(regexp.MustCompile(`^(\S+\s+){5}(.+?)\(`), 2),
+}
+
+var apacheAccessMetaExtractor = LogMetaExtractor{
+	LogLevelExtractor: logLevelImplicitInfo,
+	ProcessNameExtractor: func(syslogName string, log string) string {
+		return strings.Replace(syslogName, "_", ".", -1)
+	},
+}
+
+var apacheErrorMetaExtractor = LogMetaExtractor{
+	LogLevelExtractor: logLevelImplicitError,
+	ProcessNameExtractor: func(syslogName string, log string) string {
+		syslogName = strings.Replace(syslogName, "_", ".", -1)
+		return strings.Replace(syslogName, ".err", "", -1)
+	},
+}
+
+func NewRsyslogMetaEngine() *LogMetaEngine {
+	return &LogMetaEngine{
+		SyslogMap: map[string]*LogMetaExtractor{
```

Comment:
Should be coll to fetch that from pfconfig

Reply:
This one doesn't list files, it binds the different syslog names to their extractor so I doubt using pfconfig will be useful like for the files as this doesn't exist in our Perl codebase AFAIK and won't have any usage outside of this module

Reply:
The fact is we will need to maintain this list if we add new log files.
Why not adding a type and doing something like:

```
 logs.Element[i].Name: &LogMetaExtractor[ logs.Element[i].Type]
```

Reply:
That's because the name here is not equal to the name in pf::constants::syslog

The name in pf::constants::syslog is the name of the file. Here is the name of the syslog name which isn't part of the constants.

So I'd need to add something in the Perl constants that has no usage except for our Golang codebase here which doesn't make much sense. Plus one log file in the logs constant can have multiple syslog names (like packetfence.log) so there would be added logic to handle all of this when it has no use in Perl